aspnetcore/performance/caching/middleware.md

```diff
@@ -1,15 +1,144 @@
+
+> The Response Caching Middleware only caches 200 (OK) server responses. Any other responses, including [error pages](xref:fundamentals/error-handling), will be ignored by the middleware.
+
+## Options
+The middleware offers two options for controlling response caching.
+
+Option | Default Value
+--- | ---
+UseCaseSensitivePaths | <p>Determines if responses will be cached on case-sensitive paths.</p><p>The default value is `false`.</p>
+MaximumBodySize | <p>The largest cacheable size for the response body in bytes.</p>The default value is `64 * 1024 * 1024` [64 MB (67,108,864 bytes)].</p>
+
+The following example configures these options so that the middleware will independently cache responses on case-sensitive paths and on body response size. Configured as shown below, the middleware would independently cache the responses for `/page1` and `/PaGe1`. The middleware would also only cache responses that have a body size less than 1 MB (1,024 bytes).
+```c#
+services.AddResponseCaching(options =>
+{
+    options.UseCaseSensitivePaths = true;
+    options.MaximumBodySize = 1024;
+});
+```
+
+## `ResponseCache` attribute
+The `ResponseCache` attribute specifies the parameters necessary for setting appropriate headers for response caching. The only parameter of the `ResponseCache` attribute that strictly requires the middleware is `VaryByQueryKeys`, which does not correspond to an actual HTTP header. See [ResponseCache Attribute](response.md#responsecache-attribute) and the [ResponseCacheAttribute Class API Reference](https://docs.microsoft.com/en-us/aspnet/core/api/microsoft.aspnetcore.mvc.responsecacheattribute) for usage and a description of the parameters.
+
+## `VaryByQueryKeys` without using the `ResponseCache` attribute
+You can vary response caching with the `VaryByQueryKeys` feature without the `ResponseCache` attribute by using the `ResponseCachingFeature` directly from the `IFeatureCollection` of the `HttpContext`.
+```c#
+var responseCachingFeature = context.HttpContext.Features.Get<IResponseCachingFeature>();
+if (responseCachingFeature != null)
+{
+    responseCachingFeature.VaryByQueryKeys = new[] { "MyKey" };
+}
+```
+You can also create a custom `ResponseCachingFeature` that will apply `VaryByQueryKeys` globally to your application.
+```c#
+public void Configure(IApplicationBuilder app)
+{
+    app.UseResponseCaching();
+    
+    app.Use((context, next) =>
+    {
+        if (context.Features.Get<IResponseCachingFeature>() == null)
```

Comment:
This won't work because the feature will never be null after the response caching middleware. Also, the varybyquerykeys is only checked on the way out after the response is received so there's no point in adding it before next(). Finally, I don't think this is very applicable. Varying by query string is something that should be used for each specific resource/page. Usually there's no global query string that is available on all resources/pages in an app.